### ISSUE TITLE
"Connecting with SSH timed out" when Fish shell is used

### ISSUE THREAD

**Issue by tals** [microsoft/vscode-remote-release#2509]
Regression on #33, started after I upgraded to 1.43.0.

- VSCode Version: 1.43.0
- Local OS Version: macOS 10.14.6
- Remote OS Version: Ubuntu 18.04.3 LTS
- Remote Extension/Connection Type: SSH

Steps to Reproduce:
1. Set bash to default shell
2. Connect with VSCode remote. Works as expected.
3. Set default remote shell as Fish
2. Connecting with VSCode times out.

This was observed on two different servers

Does this issue occur when you try this locally?: N/A
Does this issue occur when you try this locally and all extensions are disabled?: N/A

```
[11:40:01.637] Log Level: 2
[11:40:01.639] remote-ssh@0.50.0
[11:40:01.639] darwin x64
[11:40:01.640] SSH Resolver called for "ssh-remote+anton", attempt 1
[11:40:01.640] SSH Resolver called for host: anton
[11:40:01.640] Setting up SSH remote "anton"
[11:40:01.643] Acquiring local install lock: /var/folders/y0/x378plh96hndqvy04th4v7880000gn/T/vscode-remote-ssh-anton-install.lock
[11:40:01.654] Looking for existing server data file at /Users/tal/Library/Application Support/Code/User/globalStorage/ms-vscode-remote.remote-ssh/vscode-ssh-host-anton-78a4c91400152c0f27ba4d363eb56d2835f9903a/data.json
[11:40:01.655] Using commit id "78a4c91400152c0f27ba4d363eb56d2835f9903a" and quality "stable" for server
[11:40:01.656] Install and start server if needed
[11:40:01.659] Checking ssh with "ssh -V"
[11:40:01.673] > OpenSSH_7.9p1, LibreSSL 2.7.3
[11:40:01.676] askpass server listening on /var/folders/y0/x378plh96hndqvy04th4v7880000gn/T/vscode-ssh-askpass-298a99f5c46843bde64ed3c630b85b9c9c6619df.sock
[11:40:01.677] Spawning local server with {"ipcHandlePath":"/var/folders/y0/x378plh96hndqvy04th4v7880000gn/T/vscode-ssh-askpass-42adc0358caea0dce208ba088ea9f67fc0396523.sock","sshCommand":"ssh","sshArgs":["-v","-T","-D","56862","-o","ConnectTimeout=30","-o","RemoteCommand=none","anton"],"dataFilePath":"/Users/tal/Library/Application Support/Code/User/globalStorage/ms-vscode-remote.remote-ssh/vscode-ssh-host-anton-78a4c91400152c0f27ba4d363eb56d2835f9903a/data.json"}
[11:40:01.677] Local server env: {"DISPLAY":"1","ELECTRON_RUN_AS_NODE":"1","SSH_ASKPASS":"/Users/tal/.vscode/extensions/ms-vscode-remote.remote-ssh-0.50.0/out/local-server/askpass.sh","VSCODE_SSH_ASKPASS_NODE":"/Applications/Visual Studio Code.app/Contents/Frameworks/Code Helper (Renderer).app/Contents/MacOS/Code Helper (Renderer)","VSCODE_SSH_ASKPASS_MAIN":"/Users/tal/.vscode/extensions/ms-vscode-remote.remote-ssh-0.50.0/out/askpass-main.js","VSCODE_SSH_ASKPASS_HANDLE":"/var/folders/y0/x378plh96hndqvy04th4v7880000gn/T/vscode-ssh-askpass-298a99f5c46843bde64ed3c630b85b9c9c6619df.sock"}
[11:40:01.679] Spawned 93505
[11:40:01.793] > local-server> Spawned ssh: 93506
[11:40:01.803] stderr> OpenSSH_7.9p1, LibreSSL 2.7.3
[11:40:03.569] stderr> debug1: Server host key: ecdsa-sha2-nistp256 SHA256:bFQr5QDzdBZHWOYB8GrdsncBvTdS47Ebd7x5ZKg/GKA
[11:40:04.998] stderr> Authenticated to <hidden> ([<hidden>]:22).
[11:40:05.920] > Welcome to Ubuntu 18.04.3 LTS (GNU/Linux 5.0.0-37-generic x86_64)
> 
>  * Documentation:  https://help.ubuntu.com
>  * Management:     https://landscape.canonical.com
>  * Support:        https://ubuntu.com/advantage
> 
>  * Latest Kubernetes 1.18 beta is now available for your laptop, NUC, cloud
>    instance or Raspberry Pi, with automatic updates to the final GA release.
> 
>      sudo snap install microk8s --channel=1.18/beta --classic
> 
>  * Multipass 1.1 adds proxy support for developers behind enterprise
>    firewalls. Rapid prototyping for cloud operations just got easier.
> 
>      https://multipass.run/
> 
>  * Canonical Livepatch is available for installation.
>    - Reduce system reboots and improve kernel security. Activate at:
>      https://ubuntu.com/livepatch
> 
> 174 packages can be updated.
> 0 updates are security updates.
> 
> Your Hardware Enablement Stack (HWE) is supported until April 2023.
> *** System restart required ***
[11:40:33.680] Terminating local server
[11:40:33.681] Resolver error: Connecting with SSH timed out
[11:40:33.685] ------




[11:40:33.687] Local server exit: 15

```

**Comment by cameronbraid** [comment]
I can confirm the same issue

**Comment by graffic** [comment]
I can confirm, using fish shell, hell broke loose.

**Comment by zmx264** [comment]
Same issue here, does not work anymore on fish shell

**Comment by qhhonx** [comment]
> I can confirm, using fish shell, hell broke loose.
> 
> At firsh ssh-remote was complaining about requiring a non-windows ssh to work. I throught that was the issue but after installing git with ssh, the problem did not go away.
> 
> I just downgraded to 0.49, as suggested by @tals, and it works again.

How can it downgrade to 0.49 ?

**Comment by graffic** [comment]
> How can it downgrade to 0.49 ?

@qhhonx In the extensions menu, search for the remote-ssh extension and in the bottom right of the ssh-remote item, you will see a cog icon. Press it and you will see a menu with an option to install other versions.

**Comment by qhhonx** [comment]
> > How can it downgrade to 0.49 ?
> 
> @qhhonx In the extensions menu, search for the remote-ssh extension and in the bottom right of the ssh-remote item, you will see a cog icon. Press it and you will see a menu with an option to install other versions.

Big thanks for you, it works again as expected.

**Comment by roblourens** [comment]
As a workaround for 0.50, you could try setting `"remote.SSH.useLocalServer": false`, and let me know whether that helps.

**Comment by zmx264** [comment]
> As a workaround for 0.50, you could try setting `"remote.SSH.useLocalServer": false`, and let me know whether that helps.

yes, setting useLocalServer to false solves the issue.

**Comment by tals** [comment]
> As a workaround for 0.50, you could try setting `"remote.SSH.useLocalServer": false`, and let me know whether that helps.

Can confirm this works

**Comment by cameronbraid** [comment]
`remote.SSH.useLocalServer=false` fixes it for me too

**Comment by michaelaye** [comment]
same problem here, was using fish for months without problems, suddenly failed (macOS 13.0.4), workaround also works here.

**Comment by roblourens** [comment]
Sorry for the trouble everyone. I can't figure out what fish is doing here to break this. I just spawn an ssh process and pipe a script into stdin, but any command sent that way is simply ignored when fish is used.

**Comment by graffic** [comment]
So I tried the other fix: `remote.SSH.useLocalServer=false` and it fixes it for me too. I guess there is already an issue about local server enabled automatically so perhaps this issue is duplicated.

**Comment by tals** [comment]
> Sorry for the trouble everyone. I can't figure out what fish is doing here to break this. I just spawn an ssh process and pipe a script into stdin, but any command sent that way is simply ignored when fish is used.

@roblourens huh. judging by the other bugs here, it seems to happen with other shells like tclsh, so maybe its not fish specific. Maybe something subtle about that bash/zsh do?

**Comment by tals** [comment]
> Sorry for the trouble everyone. I can't figure out what fish is doing here to break this. I just spawn an ssh process and pipe a script into stdin, but any command sent that way is simply ignored when fish is used.

Hey @roblourens, I think I found a workaround that allows the server to continue running.
In `getSshConnectionArgs()`, simply append `"bash"` to the ssh command, which skips fish.
I patched the extension locally and it seems to work. The integrated terminal still works.

(I couldn't find the source code on github. Would've opened a PR otherwise)

**Comment by roblourens** [comment]
Yes, that does work. However, I want to be able to attach to windows, mac, or linux remotes without knowing what platform they are ahead of time, so I can't set bash as an initial command.

Thanks for your investigation, I have a nearly identical script 😁 except that I had tried `-t` and `-ttt` (thinking of `-vvv`) neither of which work, and `-tt` actually does. It's very awkward that it causes the input and a prompt to be echoed. And I'm nervous about introducing other issues by trying it in general. It might be safer to just say that people using fish have to set a setting (either to disable local server mode or to do opt-into the `-tt` flag). I get a mess of escape characters, and I will have to strip tab characters from the install script because they trigger actual tab completion. So I'm not looking forward to the weird issues that might come up if I do this for every connection. But this is a good hint about a possible path forward, thanks!

**Comment by michaelaye** [comment]
is this related? https://github.com/fish-shell/fish-shell/issues/1439

**Comment by tals** [comment]
@roblourens ah I see.
I can think of a few options:

You could run a probe for the platform in a preflight call. This is stable, because stdin would get closed, similarly to how it is when `useLocalServer` is off.
Then, once you know that bash is available, switch over to it like in the proposed patch.

Another option (thats cleaner imo) is to rejiggle the existing bootstrap script to run a remote node agent that binds on a local port on the remote machine.
localServer then runs it together with SSH port forwarding.
You use this ssh tunnel to send commands to the remote server, which runs them.
I've use this approach before and it works really well.

**Comment by tals** [comment]
> is this related? [fish-shell/fish-shell#1439](https://github.com/fish-shell/fish-shell/issues/1439)

ha maybe. Good find!

**Comment by roblourens** [comment]
> The main issue I see with this is that the preflight would require an extra askpass for people without ssh keys, which is annoying to the user, but I think the useLocalServer=off bit has the same problem.

Many things would be simpler if I could run multiple ssh commands, but I have managed to keep it to a single ssh connection at least per window, and I don't want to backtrack on that since many people use passwords, 2fa, etc which are very annoying to enter multiple times.

> In theory once you've bootstrapped the remote machine, you can have a remote node agent that binds on a local port on the remote machine.
localServer then runs it together with SSH port forwarding, and you use this ssh tunnel to send commands to the remote server, which runs them.

Well, this is exactly what we are doing, but for every connection we still have to run a script to determine whether the remote server is already running, or needs to be updated, or which port it is listening on, etc.

**Comment by tals** [comment]
> Many things would be simpler if I could run multiple ssh commands, but I have managed to keep it to a single ssh connection at least per window, and I don't want to backtrack on that since many people use passwords, 2fa, etc which are very annoying to enter multiple times.

Makes sense. That's tricky!
My understanding is that this was the case before useLocalServer was enabled by default in 0.50 -- you had to do at least two ssh connections: the preflight one and then the bootstrap one, so backtracking from that would lose some of the advantages of localServer.

How do you feel about caching the platform information, so its amortized?

I wonder if you could reliably infer if you're dealing with windows vs linux based on stuff coming back when you're doing the ssh handshake.
For example if I run `ssh -T -v <server>`, I get some server platform information back:
```
debug1: Remote protocol version 2.0, remote software version OpenSSH_7.6p1 Ubuntu-4ubuntu0.3
debug1: match: OpenSSH_7.6p1 Ubuntu-4ubuntu0.3 pat OpenSSH_7.0*,OpenSSH_7.1*,OpenSSH_7.2*,OpenSSH_7.3*,OpenSSH_7.4*,OpenSSH_7.5*,OpenSSH_7.6*,OpenSSH_7.7* compat 0x04000002
```
Here I can tell this is linux before I issued any command.

**Comment by bobthecow** [comment]
Same issue here (macOS Catalina locally, some flavor of Ubuntu with a reasonably modern Fish remote). Downgrading fixed it, but setting `"remote.SSH.useLocalServer"` to false did not.

In fact, that was even worse. Somehow instead of just hanging, disabling it resulted in a really cool crash-loopy thing where it kept trying to connect and failing as fast as it could start, popping terminal windows open and closed.

**Comment by roblourens** [comment]
@tals, without useLocalServer we still do one connection, but we have to know the platform ahead of time, assuming Linux unless the `windowsRemotes` setting contains the current host.

Connecting multiple times the first time and caching that info is a possibility but I really want the first experience to be good. Also considered just showing a popup for the user to select the platform, then keeping that info in a setting.

@bobthecow could you share which remote.SSH settings you have set, and the log from the Remote-SSH channel?

**Comment by tals** [comment]
@roblourens got it!

**Comment by chiraag** [comment]
@roblourens I think asking for prompt on first use and then caching the information is a good compromise. As you are probably aware from the various bug-reports the current extension also breaks on (t)csh shells which are fairly common on RHEL/CentOS machines.

**Comment by roblourens** [comment]
Yeah, that issue I have not been able to reproduce however.

**Comment by chiraag** [comment]
@roblourens The failure mode is almost exactly the same as with fish. Here is a similar experiment to @tals with tcsh being the login shell on the remote machine.

```sh
$ cat > test_tcsh.sh <<EOF
bash
function test_fn {
        echo "> in test_fn"
}
test_fn
EOF

$ cat test_tcsh.sh | ssh -T remote_host
Warning: no access to tty (Bad file descriptor).
Thus no job control in this shell.
> in test_fn
function: Command not found.
}: Command not found.
test_fn: Command not found.

$ cat test_tcsh.sh | ssh -T remote_host bash
> in test_fn
```

**Comment by roblourens** [comment]
Hm, I can repro it now, I don't know what I was doing wrong before. I agree that it's probably basically the same thing.

**Comment by david50407** [comment]
Set these config works for me:
```
"remote.SSH.showLoginTerminal": true,
"remote.SSH.useLocalServer": false,
```
VScode 1.43.2
`ssh -V`: `OpenSSH_for_Windows_7.7p1, LibreSSL 2.6.5`

**Comment by benny-edlund** [comment]
I'm see this issue as well when changing my local machine from a surface laptop to a Mac.

Still connecting to the same remote host that uses tcsh.

Default shell Mac is also tcsh. 

Windows10 works fine however Mac is authenticating but breaks when seemingly sending bash commands to the remote tcsh shell

useLocalServer : false does help but you get stuck with this install terminal that seems to own the connection to the remote host

Also user experience is quite poor when needing to reconnect as you seem to get another install shell

Is it possible the issue lies with the communication with the local server rather the the remote?

**Comment by vibhoothiiaanand** [comment]
Yes I did also face same issue when fish the default shell in the remote shell(arm)
doing the temporary workaround did work fine

**Comment by fuzihaofzh** [comment]
Setting "remote.SSH.useLocalServer": false," in settings.json works for me. You can open this file by click "Edit settings.json" in Setting.

**Comment by karan173** [comment]
Any updates on a fix? Would be nice to have the functionality of remote.SSH.useLocalServer: true while using fish. Are there any local workarounds which you could suggest to make the combination work?

**Comment by fengyuentau** [comment]
Even if I uninstalled fish shell, this issue still happens.
And I have tried setting the `localServer` to `false`, but it did not help.
What I did to solve the issue at last is downgrading the `Remote-SSH` to version 0.49 following @graffic 's advice.
If you find the button `Install other version` is gray and not clickable, try uninstall and install, then install other version.

**Comment by ryan-allen** [comment]
The previous suggestions re: the config settings changes did not work for me. Changing my shell on the remote from fish to bash did work.

**Comment by jmcphers** [comment]
I also had this problem with zsh (and it was also resolved by changing to bash).

**Comment by takenaga** [comment]
I also had issue with csh, and 0.49.0 helped to solve.

Below is log from 0.49.0 and it expects bash on remote server, which @roblourens wants to avoid.
```
[22:46:27.244] Running script with connection command: ssh -T -D 54272 -o ConnectTimeout=15 <remote-host> bash
```
How about adding 'remote-shell' config parameter, and if it is specified, spawn that shell first and do the rest?

**Comment by JacobHenner** [comment]
Also experiencing this (with fish). 

Are there plans to address this? The issue has been open since March, and the suggested workaround is limiting.

I am not one to complain about bugfix timelines in open source projects. However, [this isn't an open source project](https://code.visualstudio.com/docs/remote/faq#_why-arent-the-remote-development-extensions-or-their-components-open-source), so even though I'd love to volunteer and submit a PR to fix the issue, I cannot.

**Comment by michaelaye** [comment]
@JacobHenner agree. I'm currently forced to default to bash on my remote, both ssh config and vsode workarounds do not work for me (Linux both client and host).

**Comment by MacJim** [comment]
Here's my (temporary) solution:

1. Use bash as your remote's default shell: `chsh -s /bin/bash`
2. At the end of your remote's `~/.bashrc` file, add:

```bash
# Execute `fish` in interactive mode
if [ -z "$PS1" ]; then
        echo 'This shell is not interactive. Will keep using bash to make VS Code Remote work.'
else
        echo 'This shell is interactive. Will switch to fish.'
        exec fish
fi
```

The reasoning behind this: `PS1` is unset in non-interactive shells (e.g. VSCode remote).

**Comment by tals** [comment]
> Here's my (temporary) solution:

what a great trick @MacJim !

**Comment by michaelaye** [comment]
Interestingy, @MacJim 's trick doesn't work for me!

This is the log of vscode's Remote SSH trial, and I get the time-out error again. :(

```bash
[10:37:00.149] Log Level: 2
[10:37:00.150] remote-ssh@0.55.0
[10:37:00.150] linux x64
[10:37:00.151] SSH Resolver called for "ssh-remote+luna4", attempt 1
[10:37:00.151] SSH Resolver called for host: luna4
[10:37:00.151] Setting up SSH remote "luna4"
[10:37:00.153] Acquiring local install lock: /tmp/vscode-remote-ssh-luna4-install.lock
[10:37:01.653] Starting to look for password prompt from another window
[10:37:12.172] Looking for existing server data file at /home/maye/.config/Code/User/globalStorage/ms-vscode-remote.remote-ssh/vscode-ssh-host-luna4-2af051012b66169dde0c4dfae3f5ef48f787ff69-0.55.0/data.json
[10:37:12.173] Using commit id "2af051012b66169dde0c4dfae3f5ef48f787ff69" and quality "stable" for server
[10:37:12.174] Install and start server if needed
[10:37:12.177] Checking ssh with "ssh -V"
[10:37:12.185] > OpenSSH_8.2p1 Ubuntu-4ubuntu0.1, OpenSSL 1.1.1f  31 Mar 2020

[10:37:12.188] askpass server listening on /run/user/1000/snap.code/vscode-ssh-askpass-918e431a476f0b114408fb8ce6cdf3d59fa15414.sock
[10:37:12.188] Spawning local server with {"ipcHandlePath":"/run/user/1000/snap.code/vscode-ssh-askpass-ac1a26ae380bb9f82c98a6e502a76c546b3fdf77.sock","sshCommand":"ssh","sshArgs":["-v","-T","-D","44723","-o","ConnectTimeout=15","luna4"],"dataFilePath":"/home/maye/.config/Code/User/globalStorage/ms-vscode-remote.remote-ssh/vscode-ssh-host-luna4-2af051012b66169dde0c4dfae3f5ef48f787ff69-0.55.0/data.json"}
[10:37:12.188] Local server env: {"DISPLAY":"1","ELECTRON_RUN_AS_NODE":"1","SSH_ASKPASS":"/home/maye/.vscode/extensions/ms-vscode-remote.remote-ssh-0.55.0/out/local-server/askpass.sh","VSCODE_SSH_ASKPASS_NODE":"/snap/code/46/usr/share/code/code","VSCODE_SSH_ASKPASS_MAIN":"/home/maye/.vscode/extensions/ms-vscode-remote.remote-ssh-0.55.0/out/askpass-main.js","VSCODE_SSH_ASKPASS_HANDLE":"/run/user/1000/snap.code/vscode-ssh-askpass-918e431a476f0b114408fb8ce6cdf3d59fa15414.sock"}
[10:37:12.191] Spawned 7179
[10:37:12.246] > local-server> Spawned ssh: 7187
[10:37:12.248] stderr> OpenSSH_8.2p1 Ubuntu-4ubuntu0.1, OpenSSL 1.1.1f  31 Mar 2020
[10:37:12.863] > This shell is interactive. Will switch to fish.
```

**Comment by MacJim** [comment]
> Interestingy, @MacJim 's trick doesn't work for me!

Hmm. @michaelaye Did you modify`PS1` in `.bashrc`?

Maybe let's use the `-` special parameter instead. It contains `i` when the shell is interactive.

```bash
if [[ "$-" == *"i"* ]]; then
        exec fish
fi
```

**Comment by michaelaye** [comment]
> > Interestingy, @MacJim 's trick doesn't work for me!
> 
> Hmm. @michaelaye Did you modify`PS1` in `.bashrc`?
> 
I think so?
```bash
()[maye@luna4 ~]$ echo $PS1
($(vcprompt -f %b%m))\[\033]0;\u:${NEW_PWD}\007\]\[\033[1;30m\][\[\033[0;37m\]\u\[\033[0;30m\]@\[\033[0;37m\]\h \[\033[1;34m\]${NEW_PWD}\[\033[1;30m\]]\[\033[0;37m\]\$ \[\033[0m\]
```

> Maybe let's use the `-` special parameter instead. It contains `i` when the shell is interactive.
> 
> ```shell
> if [[ "$-" == *"i"* ]]; then
>         exec fish
> fi
> ```
Using that, I don't get fish when I log in via ssh. :(

**Comment by mdelaney** [comment]
I found a way to get this working without changing the shell for the remote user. Simply enable the setting `Remote Server Listen On Socket`.

NOTE: I've only tried this with fish shell, my remote machine is Ubuntu.

Of course, this assumes you are able/willing to set `AllowStreamLocalForwarding` on the remote SSH server.

**Comment by yatamlite** [comment]
My default shell is csh as well, and using version 0.49.0 was the only solution that worked for me since I couldn't change my shell using chsh or usermod.

Thanks a ton!

**Comment by dovahcrow** [comment]
Thank you @mdelaney ! This solution works for me for ubuntu 20.04 with fish.

**Comment by asetty** [comment]
had to change my default shell back to bash on the remote shell for now to unblock myself after getting excited about my new fish env. would be great to get this fixed.

Using version 0.49.0 as some others did also didn't work for me on an ARM machine.

**Comment by AlekSi** [comment]
> workaround: in preferences, set `remote.SSH.useLocalServer` to `false`

That workaround worked for me like a charm

**Comment by tatsy** [comment]
My temporal solution:
* Use Byobu (I used tmux backend, but screen backend should be OK as well)
* Set **Byobu's** default shell as `fish` (configure `$HOME/.byobu/.tmux.conf`)
* Set **sytem's** default shell as `zsh`

Then, I can use vscode SSH through `zsh`, and can also use `fish` through Byobu when I want to do something with CLI.

**Comment by gioipv** [comment]
u can set the ssh time out is more than current time. It work for me.

**Comment by arturasmckwcz** [comment]
> As a workaround for 0.50, you could try setting `"remote.SSH.useLocalServer": false`, and let me know whether that helps.

It works, thank you!

**Comment by septatrix** [comment]
It seems like a great idea to make the ssh part robust by always using `/bin/sh`

**Comment by septatrix** [comment]
PS: This is the most upvoted bug and in the top 10 upvoted issues. Many people probably agree with me that we would wish to see this tackled in the next milestone (April)

**Comment by roblourens** [comment]
To reiterate, there is a simple workaround by disabling `useLocalServer`. But coming from some other changes, here is another workaround that will let you keep `useLocalServer` turned on (it can be faster to connect and can help with some other issues). This will currently only work in the **Nightly** build of Remote-SSH. Set this setting:

```
"remote.SSH.remotePlatform": {
        "yourhostname": "linux"
}
```

And it will add `bash` to the connection args, to skip the remote shell.

**Comment by leifwalsh** [comment]
A recent nightly update caused this to start happening for me, when it didn't happen in the past. I do have `remotePlatform` set to `linux` for this host, and I can see that the command it's trying to run does have `bash` at the end of it. If I run that outside vscode, it does connect and I can run commands, so I'm not sure what vscode is waiting for that's timing out.

**Comment by tejasvi** [comment]
I can confirm @ leifwalsh . It does not work on vscode insiders. I tried nightly version of extension as well. Remote's default shell is `bash`.

<details>
<summary>VSCode version</summary>

```
Version: 1.61.0-insider (user setup)
Commit: aa93eefe550167de3ca29aaa3d06f721a47594a3
Date: 2021-09-09T05:15:19.597Z
Electron: 13.1.8
Chrome: 91.0.4472.164
Node.js: 14.16.0
V8: 9.1.269.39-electron.0
OS: Windows_NT x64 10.0.19042
```

</details>

<details>
<summary>Extension logs</summary>

```
[09:08:51.140] Log Level: 2
[09:08:51.149] remote-ssh-nightly@2021.9.12420
[09:08:51.149] win32 x64
[09:08:51.150] SSH Resolver called for "ssh-remote+7b22686f73744e616d65223a2268747470733a2f2f756e61626c652d626c696e6b2d6c617572656e2d6d696e742e747279636c6f7564666c6172652e636f6d222c2275736572223a227473696e6768746f227d", attempt 1
[09:08:51.151] "remote.SSH.useLocalServer": false
[09:08:51.151] "remote.SSH.showLoginTerminal": false
[09:08:51.152] "remote.SSH.remotePlatform": {"xxxxxxxxxxxxxxxxxxxx":"linux"}
[09:08:51.152] "remote.SSH.path": undefined
[09:08:51.152] "remote.SSH.configFile": undefined
[09:08:51.153] "remote.SSH.useFlock": true
[09:08:51.153] "remote.SSH.lockfilesInTmp": false
[09:08:51.153] "remote.SSH.localServerDownload": auto
[09:08:51.154] "remote.SSH.remoteServerListenOnSocket": false
[09:08:51.154] "remote.SSH.showLoginTerminal": false
[09:08:51.154] "remote.SSH.defaultExtensions": []
[09:08:51.155] "remote.SSH.loglevel": 2
[09:08:51.155] SSH Resolver called for host: xxxxxx@xxxxxxxxxxxxxxxxxxxx
[09:08:51.155] Setting up SSH remote "xxxxxxxxxxxxxxxxxxxx"
[09:08:51.217] Using commit id "aa93eefe550167de3ca29aaa3d06f721a47594a3" and quality "insider" for server
[09:08:51.237] Install and start server if needed
[09:08:52.535] Checking ssh with "ssh -V"
[09:08:52.707] > O
[09:08:52.708] > p
[09:08:52.708] > e
[09:08:52.709] > n
[09:08:52.709] > S
[09:08:52.724] > S
[09:08:52.724] > H
[09:08:52.725] > _
[09:08:52.725] > f
[09:08:52.726] > o
[09:08:52.726] > r
[09:08:52.726] > _
[09:08:52.727] > W
[09:08:52.727] > i
[09:08:52.728] > n
[09:08:52.728] > d
[09:08:52.732] > o
[09:08:52.735] > w
[09:08:52.736] > s
[09:08:52.736] > _
[09:08:52.737] > 8
[09:08:52.737] > .
[09:08:52.737] > 1
[09:08:52.738] > p
[09:08:52.738] > 1
[09:08:52.739] > ,
[09:08:52.739] >  
[09:08:52.740] > L
[09:08:52.753] > i
[09:08:52.753] > b
[09:08:52.754] > r
[09:08:52.754] > e
[09:08:52.754] > S
[09:08:52.755] > S
[09:08:52.755] > L
[09:08:52.756] >  
[09:08:52.756] > 3
[09:08:52.757] > .
[09:08:52.757] > 0
[09:08:52.757] > .
[09:08:52.758] > 2
[09:08:52.758] > 

[09:08:52.777] Running script with connection command: ssh -T -D 53759 "xxxxxx@xxxxxxxxxxxxxxxxxxxx" bash
[09:08:52.787] Terminal shell path: C:\WINDOWS\System32\cmd.exe
[09:09:09.802] Resolver error: Error: Connecting with SSH timed out
	at Function.Timeout (c:\Users\User\.vscode-insiders\extensions\ms-vscode-remote.remote-ssh-nightly-2021.9.12420\out\extension.js:1:397167)
	at Timeout._onTimeout (c:\Users\User\.vscode-insiders\extensions\ms-vscode-remote.remote-ssh-nightly-2021.9.12420\out\extension.js:1:488075)
	at listOnTimeout (internal/timers.js:554:17)
	at processTimers (internal/timers.js:497:7)
[09:09:09.839] -----
```

</details>

Interestingly running `ssh -T -D 53759 "xxxxxx@xxxxxxxxxxxxxxxxxxxx" bash` outside vscode works fine.

**Comment by leifwalsh** [comment]
Latest insiders fixed this for me, FYI

**Comment by jamesrburgess** [comment]
Why does this bug keep getting closed? It still does it. Simple case, latest vscode as of today 2021/11/20 

```
Version: 1.62.3
Commit: ccbaa2d27e38e5afa3e5c21c1c7bef4657064247
Date: 2021-11-17T07:59:13.865Z (3 days ago)
Electron: 13.5.2
Chrome: 91.0.4472.164
Node.js: 14.16.0
V8: 9.1.269.39-electron.0
OS: Darwin x64 19.6.0
```

start on a mac, ssh to a RHEL7.9 machine with your default shell set to tcsh there: 
```
[14:33:50.347] stderr> OpenSSH_8.1p1, LibreSSL 2.7.3
[14:33:50.443] stderr> debug1: Server host key: ecdsa-sha2-nistp256 SHA256:thDT7iAjqI1jK/WKhEitbPrfgmnlmntN6veKP+196GI
[14:33:50.642] stderr> Authenticated to asti ([138.72.45.9]:22).
[14:33:50.797] > Warning: no access to tty (Inappropriate ioctl for device).
> Thus no job control in this shell.
[14:34:03.607] > Hello, my default shell on asti is tcsh
[14:34:04.570] > ready: c81714702934
[14:34:04.604] > Linux 3.10.0-1160.42.2.el7.x86_64 #1 SMP Tue Aug 31 20:15:00 UTC 2021
[14:34:04.605] Platform: linux
[14:34:04.645] stderr> bash: line 1: syntax error near unexpected token `fi'
[14:34:04.645] stderr> bash: line 1: `	fi'
[14:34:04.646] stderr> function: Command not found.
[14:34:04.647] > c81714702934: running
[14:34:04.647] stderr> Bad : modifier in $ '-'.
[14:34:04.652] stderr> fi: Command not found.
[14:34:04.652] stderr> found_glibc: Undefined variable.
[14:34:04.663] stderr> OAD: Command not found.
[14:34:04.668] stderr> then: Command not found.
[14:34:04.673] > Already attempted local download, failing
[14:34:04.678] stderr> fail_with_exitcode: Command not found.
[14:34:04.678] stderr> ALLOW_CLIENT_DOWNLOAD: Undefined variable.
[14:34:04.684] stderr> Unmatched '`'.
```

**Comment by DougTownsend** [comment]
I have the same problem trying to ssh into a server using tcsh

> [21:57:44.853] stderr> Authenticated to login.hpc.ncsu.edu ([152.7.96.212]:22).
[21:57:45.420] > Success. Logging you in...
[21:57:45.432] > Warning: no access to tty (Bad file descriptor).
> Thus no job control in this shell.
[21:57:45.771] > ready: 664a1a998294
[21:57:45.798] > Linux 3.10.0-1160.45.1.el7.x86_64 #1 SMP Wed Oct 13 17:20:51 UTC 2021
[21:57:45.799] Platform: linux
[21:57:45.847] stderr> bash: line 1: syntax error near unexpected token \`fi'
[21:57:45.847] stderr> bash: line 1: \`	fi'
[21:57:45.848] stderr> function: Command not found.
[21:57:45.848] stderr> Bad : modifier in $ (-).
[21:57:45.848] > 664a1a998294: running
[21:57:45.849] stderr> fi: Command not found.
[21:57:45.849] stderr> found_glibc: Undefined variable.
[21:57:45.852] stderr> OAD: Command not found.
[21:57:45.852] stderr> then: Command not found.
[21:57:45.853] > Already attempted local download, failing
[21:57:45.854] stderr> fail_with_exitcode: Command not found.
[21:57:45.868] stderr> ALLOW_CLIENT_DOWNLOAD: Undefined variable.
[21:57:45.868] stderr> Unmatched \`.

**Comment by AurevoirXavier** [comment]
I'm on zsh.
I encounter this error today.

**Comment by akshansh237** [comment]
Just encountered it. Remote-ssh extension updated to 0.74 today and it broke it.
downgrade to 0.70 helps. i am using this setting
```
    "remote.SSH.remotePlatform": {

        <remote-hostname: "linux"
    },

**Comment by tanhakabir** [comment]
@akshansh237 in `0.74` are you seeing issues with that setting being incorrect? Do you have some logs?

**Comment by shepherdjerred** [comment]
> Just encountered it. Remote-ssh extension updated to 0.74 today and it broke it. downgrade to 0.70 helps. i am using this setting
> 
> ```
>     "remote.SSH.remotePlatform": {
> 
>         <remote-hostname>: "linux"
>     },
> ```

Happened to me today as well. I set `remote.SSH.useLocalServer` to `false` and it started working for me.

**Comment by tanhakabir** [comment]
If you set the default shell to `bash` do you still have the same issue? Are you all using `zsh`?

**Comment by shepherdjerred** [comment]
I normally use `fish`. I was able to connect successfully after setting my login shell to `bash`. After I was able to connect using `bash` I set `remote.SSH.useLocalServer` to `false`. I then set my login shell to `/bin/fish` and was able to connect as normal.

**Comment by tanhakabir** [comment]
@shepherdjerred  were you able to use `fish` in the past? I always reproduced this timeout issue when using `fish` as my default shell. I'm surprised you were able to use it before

**Comment by shepherdjerred** [comment]
Yes, I was able to do it by configuring `remote.SSH.remotePlatform`. It wasn't until today that something changed.

**Comment by akshansh237** [comment]
> @akshansh237 in `0.74` are you seeing issues with that setting being incorrect? Do you have some logs?

Issue with ssh extension: if remote-server doesn't have bash as default shell, ssh fails throwing multiple incorrect syntax errors. in my case its csh.
Workaround used: add below config for each-host. i dont remember which thread i got this from, but it has worked for me for last 1 year.
```
    "remote.SSH.remotePlatform": {

        "akshanja2": "linux",
        "akshanja": "linux",
        "akshanja3": "linux",
        "akshanja4": "linux"
    },
```
Issue seen today after version upgrade to 0.74: ssh failing again due to remote shell complaining about synatx errors.
Workaround: downgraded it to 0.7

**Comment by jamesrburgess** [comment]
when this first happened to me (over a year ago) setting the remotePlatform to linux worked for a little while. Then that stopped working and I found the other workaround of setting useLocalServer to false. I'm remoting into a linux box where my shell is tcsh. The error logs are clearly errors coming from bash though. It looks like bash is trying to run a mangled script, "fi" is a normal bash command for the end of an 'if' block. It is unfortunate the log does not print out the script that is being run. Is my login shell being sent a script to be run that in turn makes a mangled script which it then tries to run with bash? That's what it looks like anyway.

**Comment by oskarkook** [comment]
I also used to be able to work around it with the `remote.SSH.remotePlatform` setting, while using fish. It stopped working yesterday after an update.

**Comment by oskarkook** [comment]
Works for me :), thanks!

**Comment by shepherdjerred** [comment]
> Added back the workaround to enable fish shells! You can try it out in the pre-release version of the extension (version `0.75.2022021903`). And have your remote in the `remote.SSH.remotePlatform` setting.
> 
> Please let me know if this works for you all

Should `remote.SSH.useLocalServer` be set to `true` or `false`?

**Comment by tanhakabir** [comment]
@shepherdjerred `remote.SSH.useLocalServer` should be set to `true`

**Comment by michaelaye** [comment]
> @shepherdjerred `remote.SSH.useLocalServer` should be set to `true`

So, but if it's a local server, why does `remote.SSH.remotePlatform` still need to be set? Just trying to understand what's happening.

**Comment by akshansh237** [comment]
> Added back the workaround to enable fish shells! You can try it out in the pre-release version of the extension (version `0.75.2022021903`). And have your remote in the `remote.SSH.remotePlatform` setting.
> 
> Please let me know if this works for you all

Thanks for fixing this. It works for me. i havent set "remote.SSH.remotePlatform" and have below setting:
```
    "remote.SSH.remotePlatform": {

        <hostname1>: "linux",
        <hostname2>: "linux"
    },
```
this is what used to work till 0.7 and broke in 0.74 and works in pre-release version 0.75.2022022115. Thanks @tanhakabir

**Comment by tanhakabir** [comment]
@michaelaye we still have a bug connecting to remotes with fish shells as their default shell. Using the `remotePlatform` setting we added a work around to make the connection work. It's not ideal but it works. 

The bug specifically has to do with what seems that Fish shells don't let us pipe in scripts after connection unless we connect with a command like `ssh your_host bash`. However, we want to avoid using "bash" since it prevents usage of RemoteCommand. For now it's not clear what the best way to accommodate both RemoteCommands and Fish shells. The `remotePlatform` setting gives me a workaround where if I see you have that set I add "bash" to the connection command and disable RemoteCommand

**Comment by septatrix** [comment]
> The bug specifically has to do with what seems that Fish shells don't let us pipe in scripts after connection unless we connect with a command like `ssh your_host bash`. However, we want to avoid using "bash" since it prevents usage of RemoteCommand. For now it's not clear what the best way to accommodate both RemoteCommands and Fish shells. The `remotePlatform` setting gives me a workaround where if I see you have that set I add "bash" to the connection command and disable RemoteCommand

I honestly would never expect `RemoteCommand` to work with this. It's correct use cases are miniscule and it prevents most stuff from working (like passing a single command or even ftp). Therefore I would have expected that this extension always suppresses `RemoteCommand` to guarantee proper functionality as it could do some unexpected stuff (change directory, start tmux etc).
The ratio of people using `RemoteCommand` to people using a "rare" shell like fish is probably very small. Even more so when excluding `RemoteCommand`s which would not affect this extension or even harm it. I honestly cannot think of a single reason why adding `-o Remote command=none` should not be the default!

Maybe you collect some telemetry which tells you more about this but if that is the case I would be very eager to see the data.

**Comment by sterni1971** [comment]
can confirm. still broken ubuntu22.04 and fish

**Comment by manasmgkar** [comment]
Still broken in 
Local OS Version: windows 10 build 19043
Remote OS Version: ubuntu 18.04 LTS

**Comment by septatrix** [comment]
After looking a bit into it, it seems like https://github.com/fish-shell/fish-shell/issues/6967 is the underlying incompatibility. Fish only starts executing commands from pipes once the pipe is closed.

**Comment by manasmgkar** [comment]
for me `Remote server listen on socket = True` and setting `use local server = false` worked for 
Host os:  windows 10 build 19043
remote os : ubuntu 18.04 lts

**Comment by sb6778** [comment]
fix please

**Comment by rbavery** [comment]
This is also a problem when setting up remote connections from WSL to an EC2 instance

**Comment by shepherdjerred** [comment]
I no longer have this issue with fish. It was fixed this time last year. Set `remote.SSH.remotePlatform` to `linux`.

**Comment by RoversCode** [comment]
Today, I also encountered the vscode remote ssh time out issue. The most elegant solution is to click on 'Kill Local Connection Server For Host' on the vscode command search page. This may be useful because the previous SSH connection was not properly terminated?

**Comment by dmikushin** [comment]
I feel sorry for SW devs who could only express their technical knowledge, without producing any useful result, and for users who are copy-pasting the same delusions all around dozens of times. This is not going anywhere.

What a solution to this issue has to give is an ability to use fish for remote SSH terminal, jointly with a working VSCode remote not throwing stupid errors. If this is what you are looking for, here is the solution:

1. Set user's shell to `/usr/bin/bash` in `/etc/passwd` your remote machine. Yes, not fish but bash. We remember that we want fish.
2. On your local machine from where VSCode is supposed to connect, create a `~/.ssh/config` entry:

```
Host remote
      User marcusmae
      Hostname localhost
      Port 64039
      IdentityFile ~/.ssh/keys/id_ed25519.remote
      IdentitiesOnly yes
      ForwardAgent yes
      ProxyJump gateway
      AddKeysToAgent yes
      RequestTTY yes
      RemoteCommand fish
```

I'm copy-pasting my own config, yours could be different. One important line is `RemoteCommand fish`, which says that when you login from terminal, you execute fish. Also you need `RequestTTY yes`.

Why this also works for VSCode? It works because VSCode overrides RemoteCommand from the command line by issuing `-o RemoteCommand none`. So VSCode gets whatever is knows how to work with, while we still get our lovely fish.

Side-effects are that there is no motd (greeting message), and SHELL env var points to bash. Probably both could be addressed with some more scripting, but important is that VSCode is defeated.

**Comment by mggluscevic59** [comment]
> **workaround**: in preferences, set `remote.SSH.useLocalServer` to `true` and `remote.SSH.remotePlatform` like so:
> 
> ```
> "remote.SSH.remotePlatform": {
>         <hostname1>: "linux",
>         <hostname2>: "linux"
>     },
> ```
> 
> ~**workaround**: Downgrading this extension to 0.49.0 fixes the issue (although it appears to break the "recent connections" UI). Extensions menu -> Right-click on "Remote - SSH" -> "Install Another Version" -> 0.49.0~
> 
> This is a regression on #33, started after I upgraded to VScode 1.43.0/extension 0.50.0.
> 
> * VSCode Version: 1.43.0
> * Local OS Version: macOS 10.14.6
> * Remote OS Version: Ubuntu 18.04.3 LTS
> * Remote Extension/Connection Type: SSH
> 
> Steps to Reproduce:
> 
> 1. Set bash to default shell
> 2. Connect with VSCode remote. Works as expected.
> 3. Set default remote shell as Fish
> 4. Connecting with VSCode times out.
> 
> This was observed on two different servers
> 
> Does this issue occur when you try this locally?: N/A Does this issue occur when you try this locally and all extensions are disabled?: N/A
> 
> ```
> [11:40:01.637] Log Level: 2
> [11:40:01.639] remote-ssh@0.50.0
> [11:40:01.639] darwin x64
> [11:40:01.640] SSH Resolver called for "ssh-remote+anton", attempt 1
> [11:40:01.640] SSH Resolver called for host: anton
> [11:40:01.640] Setting up SSH remote "anton"
> [11:40:01.643] Acquiring local install lock: /var/folders/y0/x378plh96hndqvy04th4v7880000gn/T/vscode-remote-ssh-anton-install.lock
> [11:40:01.654] Looking for existing server data file at /Users/tal/Library/Application Support/Code/User/globalStorage/ms-vscode-remote.remote-ssh/vscode-ssh-host-anton-78a4c91400152c0f27ba4d363eb56d2835f9903a/data.json
> [11:40:01.655] Using commit id "78a4c91400152c0f27ba4d363eb56d2835f9903a" and quality "stable" for server
> [11:40:01.656] Install and start server if needed
> [11:40:01.659] Checking ssh with "ssh -V"
> [11:40:01.673] > OpenSSH_7.9p1, LibreSSL 2.7.3
> [11:40:01.676] askpass server listening on /var/folders/y0/x378plh96hndqvy04th4v7880000gn/T/vscode-ssh-askpass-298a99f5c46843bde64ed3c630b85b9c9c6619df.sock
> [11:40:01.677] Spawning local server with {"ipcHandlePath":"/var/folders/y0/x378plh96hndqvy04th4v7880000gn/T/vscode-ssh-askpass-42adc0358caea0dce208ba088ea9f67fc0396523.sock","sshCommand":"ssh","sshArgs":["-v","-T","-D","56862","-o","ConnectTimeout=30","-o","RemoteCommand=none","anton"],"dataFilePath":"/Users/tal/Library/Application Support/Code/User/globalStorage/ms-vscode-remote.remote-ssh/vscode-ssh-host-anton-78a4c91400152c0f27ba4d363eb56d2835f9903a/data.json"}
> [11:40:01.677] Local server env: {"DISPLAY":"1","ELECTRON_RUN_AS_NODE":"1","SSH_ASKPASS":"/Users/tal/.vscode/extensions/ms-vscode-remote.remote-ssh-0.50.0/out/local-server/askpass.sh","VSCODE_SSH_ASKPASS_NODE":"/Applications/Visual Studio Code.app/Contents/Frameworks/Code Helper (Renderer).app/Contents/MacOS/Code Helper (Renderer)","VSCODE_SSH_ASKPASS_MAIN":"/Users/tal/.vscode/extensions/ms-vscode-remote.remote-ssh-0.50.0/out/askpass-main.js","VSCODE_SSH_ASKPASS_HANDLE":"/var/folders/y0/x378plh96hndqvy04th4v7880000gn/T/vscode-ssh-askpass-298a99f5c46843bde64ed3c630b85b9c9c6619df.sock"}
> [11:40:01.679] Spawned 93505
> [11:40:01.793] > local-server> Spawned ssh: 93506
> [11:40:01.803] stderr> OpenSSH_7.9p1, LibreSSL 2.7.3
> [11:40:03.569] stderr> debug1: Server host key: ecdsa-sha2-nistp256 SHA256:bFQr5QDzdBZHWOYB8GrdsncBvTdS47Ebd7x5ZKg/GKA
> [11:40:04.998] stderr> Authenticated to <hidden> ([<hidden>]:22).
> [11:40:05.920] > Welcome to Ubuntu 18.04.3 LTS (GNU/Linux 5.0.0-37-generic x86_64)
> > 
> >  * Documentation:  https://help.ubuntu.com
> >  * Management:     https://landscape.canonical.com
> >  * Support:        https://ubuntu.com/advantage
> > 
> >  * Latest Kubernetes 1.18 beta is now available for your laptop, NUC, cloud
> >    instance or Raspberry Pi, with automatic updates to the final GA release.
> > 
> >      sudo snap install microk8s --channel=1.18/beta --classic
> > 
> >  * Multipass 1.1 adds proxy support for developers behind enterprise
> >    firewalls. Rapid prototyping for cloud operations just got easier.
> > 
> >      https://multipass.run/
> > 
> >  * Canonical Livepatch is available for installation.
> >    - Reduce system reboots and improve kernel security. Activate at:
> >      https://ubuntu.com/livepatch
> > 
> > 174 packages can be updated.
> > 0 updates are security updates.
> > 
> > Your Hardware Enablement Stack (HWE) is supported until April 2023.
> > *** System restart required ***
> [11:40:33.680] Terminating local server
> [11:40:33.681] Resolver error: Connecting with SSH timed out
> [11:40:33.685] ------
> 
> 
> 
> 
> [11:40:33.687] Local server exit: 15
> ```

In my case:
`"lima-default": "linux"`

**Comment by DeflateAwning** [comment]
Why can't this plugin send `RemoteCommand bash` if it's a well-known bug that `fish` shell breaks it. Seems like an obvious fix.

**Comment by ehartford** [comment]
what is the current best workaround?  I see several and can't figure out what it the right one
is there any effort to resolve this issue for good?

**Comment by dmikushin** [comment]
For myself I've ended up having a special flavor of dropbear running locally, which turns incoming SSH connection into bash session, iff the username has _vscode suffix: https://github.com/dmikushin/dropbear/commit/864d93e937b559da350c5042569019d28367ef73

**Comment by pwr22** [comment]
Still a problem for me too.

**Comment by ehartford** [comment]
Also really not an easy issue to identify - I had to guess after several days of troubleshooting that it was related to fish shell before I found this issue.  There was no hints of the cause, it simply timed out

**Comment by michaelaye** [comment]
Funnily enough, vscodium with the remote extension works, cursor.sh with same extension doesn't.

**Comment by dmikushin** [comment]
This extension is not opensource, right? A great example of how hiding the source code leads to just endless 4-year long discussions.

**Comment by michaelaye** [comment]
> Funnily enough, vscodium with the remote extension works, cursor.sh with same extension doesn't.

update: On Cursor, all that was needed was to set the remote platform to linux for the host I connect to. I am coming from a Mac, so cursor was wrongly assuming that I connect to a mac, at least that's was the first setting popping up when i defined the new remote platform item in the settings.

**Comment by solariz** [comment]
Greetings from 2025; Just had to re-setup my system and unfortunately this is still an issue, need to set remote.SSH.useLocalServer=false to have it working with remote default fish shell.

**Comment by ggladilov** [comment]
Faced the same issue here

**Comment by ODtian** [comment]
Same issue

**Comment by F1yingWhite** [comment]
Same issue

**Comment by MovFish** [comment]
Same issue, it doesn't happen on Windows but happens on Arch Linux.

**Comment by antomanc** [comment]
same issue

**Comment by rafaelsms** [comment]
Please, stop commenting "Same issue" or "+1" or similar! If you have nothing _new_ to add, avoid commenting.

A lot of us receives emails for this because we have interest in the issue. If you just want to "make pressure" to have this fixed, just add a reaction to other's comments, to the issue description or subscribe to it. Adding useless comments might cause it to get locked (rightfully), which inhibits useful discussion and feedback.

**Comment by NightWatcher314** [comment]
In my case, set UseLocalServer not working.
So, I install a clean zsh, add only "fish" to ~/.zshrc, then run "chsh -s /usr/bin/zsh".
It works.

**Comment by lesca** [comment]
+1
nobody gonna fix this after 5 years 😢 
my workaround is use `bash` and type `fish` to switch shell

**Comment by dmikushin** [comment]
Hey, I've found one particularly neat solution:

1. Get a separate SSH key for VS Code.
2. In the `authorized_keys` on the server advertising the public part of the SSH key, prepend it with a `command=` part saying that you wish to execute bash instead of whatever the default user shell is:

```
command="/bin/bash" ssh-ed25519 AAAA...
```

Tip-top. Now we have a solution, and Microsoft can close this, because they can't understand what's going on here anyways.

**Comment by marcoce7** [comment]
In my case, to connect to a machine that uses fish as default shell it is enough to add the hostname that I want to connect to to `remote.SSH.remotePlatform` in settings with value `linux`, as in the opening comment of this issue, <https://github.com/microsoft/vscode-remote-release/issues/2509#issue-578458131>. In the JSON, this is something like

```
"remote.SSH.remotePlatform": {
  "hostname-with-fish-shell": "linux"
}
```

It works with `remote.SSH.useLocalServer` set to `true`, which is the default.
I think it simply skips some initialization script to determine the platform that breaks when the shell is not bash?

This way, I do not need to edit `~/.ssh/config` of the local machine to add a special host with a different remote command, which I think requires disabling `remote.SSH.useLocalServer` to work. Nor I need to use a separate SSH key.

**Comment by JHXs** [comment]
> In my case, to connect to a machine that uses fish as default shell it is enough to add the hostname that I want to connect to to `remote.SSH.remotePlatform` in settings with value `linux`, as in the opening comment of this issue, [#2509 (comment)](https://github.com/microsoft/vscode-remote-release/issues/2509#issue-578458131). In the JSON, this is something like
> 
> ```
> "remote.SSH.remotePlatform": {
>   "hostname-with-fish-shell": "linux"
> }
> ```
> 
> It works with `remote.SSH.useLocalServer` set to `true`, which is the default. I think it simply skips some initialization script to determine the platform that breaks when the shell is not bash?
> 
> This way, I do not need to edit `~/.ssh/config` of the local machine to add a special host with a different remote command, which I think requires disabling `remote.SSH.useLocalServer` to work. Nor I need to use a separate SSH key.

Thanks, This fixes it for me.

**Comment by Honahec** [comment]
> In my case, to connect to a machine that uses fish as default shell it is enough to add the hostname that I want to connect to to `remote.SSH.remotePlatform` in settings with value `linux`, as in the opening comment of this issue, [#2509 (comment)](https://github.com/microsoft/vscode-remote-release/issues/2509#issue-578458131). In the JSON, this is something like
> 
> ```
> "remote.SSH.remotePlatform": {
>   "hostname-with-fish-shell": "linux"
> }
> ```
> 
> It works with `remote.SSH.useLocalServer` set to `true`, which is the default. I think it simply skips some initialization script to determine the platform that breaks when the shell is not bash?
> 
> This way, I do not need to edit `~/.ssh/config` of the local machine to add a special host with a different remote command, which I think requires disabling `remote.SSH.useLocalServer` to work. Nor I need to use a separate SSH key.

Thanks! Really helpful.

BTW, I can add `remote.SSH.remotePlatform` automatically by connecting a machine that uses fish after setting `remote.SSH.useLocalServer` to `false`.

Then, set `remote.SSH.useLocalServer` to `true`, and everything all right.

**Comment by Honahec** [comment]
Nobody gonna fix this after 6 years...
